scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +159,306 @@
+          current-db-config-ent (current-db-config-ents name)
+          current-in-mem-config (current-in-mem-configs name)]
+      (when (and current-in-mem-config (not (and cluster state-atom state-locked?-atom)))
+        (throw (ex-info "We know an in-memory config but we don't know the corresponding in-memory cluster.
+        This should never happen, since cluster-name->compute-cluster-atom should be locked"
+                        {:current-in-mem-config current-in-mem-config
+                         :cluster cluster})))
+      (when (and state-atom state-locked?-atom)
+        (reset! state-atom state)
+        (reset! state-locked?-atom state-locked?))
+      (when current-db-config-ent
+        @(d/transact
+           conn
+           [(assoc (compute-cluster-config->compute-cluster-config-ent config) :db/id (:db/id current-db-config-ent))])))
+    {:update-succeeded true}
+    (catch Throwable t
+      (log/error t "Failed to update cluster" config)
+      {:update-succeeded false :error-message (.toString t)})))
+
+; TODO make sure anything in "locking" path times out. like db calls
+; TODO do alerts in callers of this function on return value
+(defn update-compute-clusters
+  "This function allows adding or updating the current compute cluster configurations. It takes
+  in a single configuration and/or a map of configurations with cluser name as the key. Passing in a map of
+  configurations implies that these are the only known configurations, and clusters that are missing from this
+  map should be removed.
+  Only the state of an existing cluster and its configuration can be changed unless force? is set to true."
+  [conn new-config new-configs force?]
+  {:pre [(= 1 (->> [new-config new-configs] (filter some?) count))]}
+  (locking cluster-name->compute-cluster-atom
+    (let [db (d/db conn)
+          current-db-config-ents (db-config-ents db)
+          current-in-mem-configs (in-mem-configs)
+          current-db-configs (map-vals compute-cluster-config-ent->compute-cluster-config current-db-config-ents)
+          current-configs (compute-current-configs current-db-configs current-in-mem-configs)
+          new-configs' (cond-> (or new-configs current-configs) new-config (assoc (:name new-config) new-config))
+          updates (compute-config-updates db current-configs new-configs' force?)]
+      (log/info "Updating dynamic clusters." {:current-configs current-configs :new-config new-config :new-configs new-configs :force? force? :updates updates})
+      (->> updates
+           (map #(let [{:keys [valid? insert? update? config]} %]
+                   (assoc %
+                     :update-result
+                     (when valid?
+                       (cond
+                         insert? (add-new-cluster! conn config)
+                         update? (update-cluster! conn config current-db-config-ents current-in-mem-configs))))))
+           doall))))
+
+(defn get-compute-clusters
```

Comment:
How can these be created, or is this only created manually?
Specifically,  clusters-in-db  (but not in memory) : where do they come from? How are they inserted, or only inserted directly by the database?

Reply:
clusters-in-db (but not in memory)  - this  happens at startup

Reply:
but cluster configs in the database can have changes that are not yet reflected in the in-memory clusters and will not be until a restart